### PR TITLE
feat: server signer rotation support

### PR DESCRIPTION
**Pull request by d4rp4t**
## Summary

⚠️ Stacks on #237
This PR is rebased on top of migrate-regtest-denigiri — please merge #237 first, or review/merge both together. The base branch will be updated to master automatically once #237 lands.

Client-side support for arkd signer-key rotation, mirroring what was done in ts-sdk #554 and dotnet-sdk #132. When the operator rotates its signing key, existing VTXOs and boarding outputs remain under the old key. This PR lets the SDK discover, classify, and cooperatively migrate those funds to the active signer, and recover Boltz swaps that straddle a rotation.

> **Depends on #242** (X-Digest header / `refresh_server_info()` / `ServerInfoChanged`). That PR populates `server_info.deprecated_signers` at runtime — without it the fields added here are always empty and the whole feature is a no-op. The code in this PR compiles and is safe to merge independently; the feature activates once #242 lands.

## Three regimes

| Regime | Condition | Behaviour |
|---|---|---|
| **Migratable** | deprecated signer, before cutoff (or `cutoffDate == 0`) | cooperative settle to current signer via `migrate_deprecated_signer_vtxos()` |
| **Pending recovery** | deprecated signer, past cutoff, not yet expired | excluded from offchain spend and batch; counted in new `OffChainBalance::pending_recovery` bucket; unilateral exit still available |
| **Recoverable** | expired | `settle()` picks up automatically — no forfeit needed |

## Changes

**`ark-core`**
- `Info::all_server_keys()` — iterates current + all deprecated signers
- `Info::is_signer_past_cutoff_at()` — returns true when a signer's cutoff has passed; `cutoff_date == 0` means "rotate immediately" and is treated as not-yet-past-cutoff (operator still co-signs; migration is triggered separately)
- `BoardingOutput::server_pk()` — exposes the server key embedded in the boarding output script

**`ark-client`**
- `get_offchain_addresses()` and `get_boarding_addresses()` fan out over `current ∪ deprecated` signers so all funds remain visible
- `OffChainBalance` gains `pending_recovery` bucket; past-cutoff VTXOs are excluded from     `confirmed`/`pre_confirmed`
- `migrate_deprecated_signer_vtxos()` — sweeps VTXOs **and boarding outputs** under pre-cutoff deprecated signers to the current signer by joining the next batch
- `fetch_commitment_transaction_inputs()` excludes past-cutoff VTXOs and boarding outputs (operator won't co-sign the old key's forfeit path — including them would brick the batch)
- `auto_select_send_inputs()` excludes past-cutoff VTXOs from offchain sends
- `discover_keys()` generates addresses for all known server keys so user keys active before a rotation are found on seed re
- Boltz: all VHTLC reconstruction sites replaced with `reconstruct_vhtlc_for_address()`, which tries current + deprecated server keys to match the stored lockup address — swap recovery works across a rotation

**`ark-rest`**
- `DeprecatedSigner` conversion: missing `cutoffDate` in JSON now maps to `0` ("rotate immediately") instead of returning a `ConversionError` and crashing `get_info`

## Summary by CodeRabbit

* **New Features**
  * Added time-based cutoffs for server signer key rotation.
  * Introduced `pending_recovery` in offchain balance totals for assets held under deprecated keys past cutoff.
  * Expanded offchain address discovery to include current and deprecated server keys.
  * Improved migration of deprecated-signer VTXOs with cutoff-aware behavior.

* **Bug Fixes**
  * Excludes cutoff-past boarding and deprecated-signer VTXOs from batch input selection and auto-selection.
  * Updates VHTLC reconstruction to match the swap’s expected address during signer rotation.

* **Tests**
  * Added end-to-end coverage for signer rotation and deprecated-signer migration (including past/future cutoff cases).

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
## 🔍 Aggressive Code Review — Server Signer Rotation Support

**Reviewer**: Arkana  
**Verdict**: Request changes — one fund-safety regression, missing test coverage, and a few issues to address.

---

### 🔴 CRITICAL — `collect_pending_vhtlcs` error propagation regression (fund-safety)

**Files**: `ark-client/src/boltz.rs:3506–3511` and `3556–3561` (line numbers per current master; in the PR these are the `build_vhtlc_script(..., &swap.vhtlc_address)?` call sites)

**Before**: `build_vhtlc_script()` always returned a VHTLC (built with current signer key). If the resulting address didn't match the swap's stored address, the code logged a warning and `continue`d — **skipping that one swap** while processing the rest.

**After**: `build_vhtlc_script()` now calls `reconstruct_vhtlc_for_address()` internally, which iterates current + all deprecated keys. If **none** match, it returns `Err(...)`. The `?` operator propagates this error, **killing the entire `collect_pending_vhtlcs()` operation**.

**Impact**: If the server removes a deprecated signer from `deprecated_signers` (e.g., after all its VTXOs expired) but the client still has a swap record referencing that old key, **one stale swap record blocks recovery of ALL pending VHTLCs**. This is a loss-of-funds vector — active swaps with valid keys can't be recovered because of an unrelated stale record.

**Fix**: In `collect_pending_vhtlcs`, catch the `Err` from `build_vhtlc_script` per-swap and `warn` + `continue` instead of propagating:
```rust
let vhtlc = match self.build_vhtlc_script(
    swap.claim_public_key,
    swap.refund_public_key,
    swap.preimage_hash,
    &swap.timeout_block_heights,
    &swap.vhtlc_address,
) {
    Ok(v) => v,
    Err(e) => {
        tracing::warn!(swap_id = swap.id, error = %e, "VHTLC reconstruction failed, skipping");
        continue;
    }
};
```

This preserves the old "skip bad, process good" behavior while still gaining the multi-key reconstruction.

---

### 🟡 MODERATE — `migrate_deprecated_signer_vtxos` settles ALL VTXOs

**File**: `ark-client/src/lib.rs:1145–1170` (the `settle(rng).await` call)

The method detects whether migratable deprecated-signer outputs exist, but then calls `self.settle(rng)` which settles **every** unspent VTXO — current-signer included. If a user has 100 current-signer VTXOs and 1 deprecated-signer VTXO, all 101 are settled, incurring fees on 100 VTXOs that didn't need to move.

This is documented ("settle() picks up all unspent VTXOs") so it may be intentional, but:
1. Consider documenting this cost in the public API doc-comment
2. Consider whether a targeted send-to-self would be more appropriate for users with large VTXO sets

---

### 🟡 MODERATE — Zero test coverage for rotation logic

No unit tests or e2e tests cover:
- `is_signer_past_cutoff_at` — the core decision function for the entire feature
- `all_server_keys` iteration order
- `reconstruct_vhtlc_for_address` key-matching logic
- `migrate_deprecated_signer_vtxos` flow
- `pending_recovery` balance bucket accounting
- The `cutoff_date == 0` → "not past cutoff" semantics
- Edge cases: duplicate deprecated signers, empty `deprecated_signers`, past-cutoff == exactly now

At minimum, `is_signer_past_cutoff_at` and `reconstruct_vhtlc_for_address` need unit tests — they are the safety-critical decision points that determine whether VTXOs can be spent.

---

### 🟡 MODERATE — `cutoff_date == 0` means two different things

**File**: `ark-core/src/server.rs:497–501` and `ark-rest/src/conversions.rs:120`

- In `is_signer_past_cutoff_at`: `cutoff_date == 0` → NOT past cutoff (operator still co-signs)
- In `migrate_deprecated_signer_vtxos`: `cutoff_date == 0` → IS migratable (`ds.cutoff_date == 0 || ds.cutoff_date > now`)
- In `ark-rest`: missing JSON `cutoffDate` → `0` ("rotate immediately")

The semantics are consistent if you read all three together, but `0` being a sentinel value in an `i64` unix-timestamp field is fragile. A negative `cutoff_date` from a buggy server would also pass `!= 0` and `<= now`, marking the signer as past-cutoff. Consider adding a validation bound (e.g., `cutoff_date > 0` means "has a real cutoff") or using `Option<i64>` at the core level.

---

### 🟢 MINOR — Redundant `parse_sequence_number` in boltz.rs closures

**File**: `ark-client/src/boltz.rs` — all 6 `reconstruct_vhtlc_for_address` call sites

The `parse_sequence_number` calls are inside the `mk_opts` closure, meaning they execute N times (once per server key attempted). These are pure functions with identical inputs — they should be hoisted above the closure like `build_vhtlc_script` already does (lines 3437–3444 in the diff). Not a correctness issue, but wasteful.

---

### 🟢 INFO — Cross-repo impact

1. **ark-flutter-example**: The FRB bridge at `rust/src/api/ark_api.rs:48-53` wraps `OffChainBalance` with positional SSE encoding. Adding `pending_recovery` will break the bridge unless the wrapper is updated and FRB bindings regenerated. Track this.

2. **go-sdk PR #196** ("Signer Key rotation") is the parallel Go implementation. Review for API parity — Go already has `DeprecatedSignerStatus` and `reconcileDeprecatedSigners` which maps to `migrate_deprecated_signer_vtxos`.

3. No breakage in solver, bancod, fulmine, or ark — they access `OffchainBalance` through Go SDK types (separate struct, JSON-permissive).

---

### ✅ What looks good

- `reconstruct_vhtlc_for_address` is a clean abstraction — try current key first, fall back to deprecated keys. Good pattern.
- `is_signer_past_cutoff_at` correctly excludes past-cutoff VTXOs from batches and sends — including them would brick the batch (operator can't co-sign old-key forfeits).
- `discover_keys` correctly fans out over all server keys — seed recovery works across rotations.
- `get_offchain_addresses` and `get_boarding_addresses` covering deprecated signers means balances stay visible.
- The `ark-rest` defensive fix for missing `cutoffDate` prevents a crash on `get_info`.
- Boarding output filtering in `fetch_commitment_transaction_inputs` is correct — past-cutoff boarding outputs are excluded.
- The `is_excluded` closure in batch.rs correctly allows recoverable VTXOs through (no forfeit needed).

---

### ⚠️ PROTOCOL-CRITICAL FLAG

This PR modifies VTXO handling, forfeit path inclusion/exclusion, batch lifecycle inputs, and exit path logic. **Requires human review and sign-off before merge**, even after code issues are addressed.

Depends on #242 (not yet merged).

**Review by ghost — COMMENTED** [review]
## 🔍 Follow-up Review — New Commits (1f5d4160..c4ca97a5)

Four commits pushed since my last review. All address prior feedback:

### ✅ `1f5d4160` — `collect_pending_vhtlcs` error propagation fix
**Critical finding resolved.** Both submarine and reverse swap loops now `match` on `build_vhtlc_script` and `warn! + continue` on error. One stale swap record no longer kills recovery of all pending VHTLCs. Implementation matches the suggested fix exactly.

### ✅ `322002ce` — Doc comment on settle scope
`migrate_deprecated_signer_vtxos` doc now clearly states that `settle()` consolidates **all** VTXOs (current-signer included) and acknowledges the fee cost. Good.

### ✅ `faf98012` — Hoist `parse_sequence_number` out of closures
All 7 `reconstruct_vhtlc_for_address` call sites parse sequence numbers once before the closure. Clean mechanical refactor, no logic changes.

### ✅ `c4ca97a5` — Explicit `js-sys` wasm32 dependency
`js-sys` added to `ark-client/Cargo.toml` under wasm32 deps. Prevents breakage if the transitive dep chain changes.

### Remaining open items (from initial review, unchanged)

1. **Test coverage** — still zero unit tests for `is_signer_past_cutoff_at`, `reconstruct_vhtlc_for_address`, `pending_recovery` bucket, and `cutoff_date == 0` semantics. Not blocking but should be tracked.

2. **`cutoff_date == 0` sentinel** — design concern acknowledged, now documented. Acceptable for this PR.

3. **Cross-repo: ark-flutter-example FRB bridge** — `OffChainBalance::pending_recovery` will need the FRB wrapper updated. Still needs tracking.

### ⚠️ PROTOCOL-CRITICAL — human sign-off required
This PR modifies VTXO handling, forfeit path inclusion/exclusion, and batch lifecycle inputs. Requires human review before merge, even with code issues resolved.

No new issues found in the delta commits. Keeping my "changes requested" status until a human reviews — not because of remaining code defects, but because protocol-critical changes require human sign-off per policy.

**Review by ghost — APPROVED** [review]
## 🔍 Follow-up Review — New Commits (c4ca97a5..8c0dc9b8)

Two new commits since my last review. Both address prior feedback.

### ✅ `b6bd86cc` — VHTLC construction deduplication

Seven inline `reconstruct_vhtlc_for_address` call sites collapsed into `build_vhtlc_script(claim, refund, preimage_hash, timeouts, expected_address)`. The key-role mapping (`sender=refund`, `receiver=claim`) is preserved correctly at all sites, including the chain-swap variants (`server_refund_public_key` / `server_claim_public_key`). The core iteration logic is extracted into a free function `reconstruct_vhtlc_from_keys()` to enable unit testing without a `Client` instance.

Verified all 7+2 call sites — no semantic changes, purely mechanical dedup.

### ✅ `8c0dc9b8` — Unit tests

Solid coverage across all three crates:

- **`ark-core::server`**: `all_server_keys` ordering, `is_signer_past_cutoff_at` boundary conditions (zero-cutoff, exact boundary, future, past, multi-deprecated). Covers the critical edge case of `cutoff_date == 0` meaning "rotate immediately but still co-sign".
- **`ark-rest::conversions`**: `DeprecatedSigner` parsing — valid cutoff, missing cutoff defaulting to `0`, missing/invalid pubkey errors.
- **`ark-client::boltz`**: `reconstruct_vhtlc_from_keys` — single-key match, multi-key fallback to deprecated, no-match error, `mk_opts` error propagation. Plus `build_vhtlc_script_sender_is_refund_receiver_is_claim` which anchors the key-role mapping against a fixture address.

All prior critical and non-critical findings resolved. Approving.

⚠️ **Reminder**: This is protocol-critical code (VTXO handling, signer rotation, forfeit paths). Human sign-off still required before merge.

**Review by ghost — APPROVED** [review]
## 🔍 Follow-up Review — New Commits (c4ca97a5..8c0dc9b8)

Two new commits since my last review. Both address prior feedback.

### ✅ `b6bd86cc` — VHTLC construction deduplication

Seven inline `reconstruct_vhtlc_for_address` call sites collapsed into `build_vhtlc_script(claim, refund, preimage_hash, timeouts, expected_address)`. The key-role mapping (`sender=refund`, `receiver=claim`) is preserved correctly at all sites, including the chain-swap variants (`server_refund_public_key` / `server_claim_public_key`). The core iteration logic is extracted into a free function `reconstruct_vhtlc_from_keys()` to enable unit testing without a `Client` instance.

Verified all 7+2 call sites — no semantic changes, purely mechanical dedup.

### ✅ `8c0dc9b8` — Unit tests

Solid coverage across all three crates:

- **`ark-core::server`**: `all_server_keys` ordering, `is_signer_past_cutoff_at` boundary conditions (zero-cutoff, exact boundary, future, past, multi-deprecated). Covers the critical edge case of `cutoff_date == 0` meaning "rotate immediately but still co-sign".
- **`ark-rest::conversions`**: `DeprecatedSigner` parsing — valid cutoff, missing cutoff defaulting to `0`, missing/invalid pubkey errors.
- **`ark-client::boltz`**: `reconstruct_vhtlc_from_keys` — single-key match, multi-key fallback to deprecated, no-match error, `mk_opts` error propagation. Plus `build_vhtlc_script_sender_is_refund_receiver_is_claim` which anchors the key-role mapping against a fixture address.

All prior critical and non-critical findings resolved. Approving.

⚠️ **Reminder**: This is protocol-critical code (VTXO handling, signer rotation, forfeit paths). Human sign-off still required before merge.

**Review by ghost — APPROVED** [review]
## 🔍 Re-review after rebase (b03b2e4f)

Confirmed: force-push was a rebase only — same 13 commits, same content, no new changes. My previous approval stands.

All prior findings resolved:
- ✅ `collect_pending_vhtlcs` error propagation
- ✅ Doc comment on settle scope
- ✅ `parse_sequence_number` hoisted out of closures
- ✅ Explicit `js-sys` wasm32 dependency
- ✅ VHTLC construction deduplication
- ✅ Unit test coverage for critical paths

⚠️ **PROTOCOL-CRITICAL** — This PR modifies VTXO handling, signer rotation, forfeit path inclusion/exclusion, and batch lifecycle inputs. **Human sign-off required before merge.**

**Review by ghost — CHANGES_REQUESTED** [review]
## 🔒 Arkana Protocol Review — Server Signer Rotation Support

Thorough review of all protocol-critical changes across `ark-core`, `ark-rest`, `ark-client` (lib, batch, send_vtxo, boltz), e2e tests, and cross-repo impact. No critical or high-severity issues found. The design correctly partitions VTXOs into three regimes (migratable / pending_recovery / recoverable) with no gap where funds could be lost.

**Requesting changes for the two medium-severity items below. Also flagging for mandatory human sign-off — this is protocol-critical code (VTXO handling, batch input selection, forfeit path exclusion).**

---

### 🟡 MEDIUM — Inconsistent time sources for cutoff checks

`offchain_balance()` uses `unix_now()` (`std::time::SystemTime` on native, `js_sys::Date` on WASM) at `lib.rs:~1055`, while `fetch_commitment_transaction_inputs()` uses `jiff::Timestamp::now()` at `batch.rs:~1014` and `batch.rs:~1071`. These are two different time source abstractions. On native they should agree, but at the cutoff boundary a VTXO could be classified as `confirmed` in the balance yet excluded from batch inputs (or vice versa) if the two sources diverge by even one second.

**Recommendation:** Use the same time source in both places. Either `unix_now()` everywhere or `Timestamp::now()` everywhere.

### 🟡 MEDIUM — TOCTOU race in `migrate_deprecated_signer_vtxos()`

At `lib.rs:~1136`, the function captures `now = unix_now()` and uses it for the `is_pre_cutoff_deprecated` check. But `settle()` (called at `lib.rs:~1172`) internally calls `fetch_commitment_transaction_inputs()` which captures its own `now`. If the cutoff passes between the two snapshots, the VTXO will pass the migration gate but be silently excluded from batch inputs. No funds are lost (VTXOs become `pending_recovery`), but the migration reports success while doing nothing for those VTXOs.

**Recommendation:** Pass the captured `now` timestamp through to `settle()`/`fetch_commitment_transaction_inputs()` so the entire migration uses a single time snapshot.

---

### 🟢 LOW — `cutoff_date == 0` sentinel is correct but fragile

`ark-rest/src/conversions.rs:~124`: Missing `cutoffDate` defaults to `0`. This works today because `is_signer_past_cutoff_at()` has a `!= 0` guard and `migrate_deprecated_signer_vtxos()` checks `ds.cutoff_date == 0` separately. But the dual semantics are implicit — if future code checks `cutoff_date < now` without the `!= 0` guard, `0` (Unix epoch) would incorrectly read as "long past cutoff."

**Recommendation:** Add a doc comment on `DeprecatedSigner::cutoff_date` documenting the sentinel: `/// 0 means "rotate immediately" — operator still co-signs, treated as pre-cutoff.`

### 🟢 LOW — Manual VTXO selection paths don't filter past-cutoff VTXOs

`send_selection()` → `resolve_selected_send_inputs()` and `collaborative_redeem_vtxo_selection()` do not check `is_signer_past_cutoff_at`. If a user explicitly selects a past-cutoff VTXO, the transaction will be submitted but the operator will refuse to co-sign. No funds lost (server rejects it), but the error won't explain why.

**Recommendation:** Either filter or return a descriptive error when past-cutoff VTXOs are manually selected.

### 🟢 LOW — `fetch_commitment_transaction_inputs` captures `now` twice

`batch.rs:~1014` (for boarding outputs) and `batch.rs:~1071` (for VTXOs) make separate `Timestamp::now()` calls with a `list_vtxos()` network call in between. Narrow window, but boarding outputs and VTXOs could be filtered against different timestamps.

**Recommendation:** Capture `now` once at the top of the function.

---

### ✅ What looks good

- **`all_server_keys()` / `is_signer_past_cutoff_at()`** — Correct, well-tested (7 edge-case tests), ordering guarantee (current first) is maintained.
- **`OffChainBalance::pending_recovery`** — Correctly included in `total()`. No double-counting: `spendable_offchain()` and `recoverable()` are disjoint sets.
- **VHTLC reconstruction (`reconstruct_vhtlc_from_keys`)** — Address-match check was not removed, it was consolidated into the reconstruction loop. All 9 call sites reviewed — semantics preserved, key-role mapping (sender=refund, receiver=claim) consistent everywhere. Error propagation intact.
- **`BoardingOutput::server_pk()`** — Safe accessor, x-only public key, no sensitive data.
- **Batch input exclusion** — `is_excluded` closure correctly skips past-cutoff non-recoverable VTXOs. Recoverable VTXOs pass through (no forfeit needed). This prevents the most dangerous failure: bricking a round by including VTXOs the operator can't co-sign.
- **Cross-repo impact** — All new APIs are additive. `OffChainBalance` is not deserialized by external consumers. No breaking changes.
- **E2e tests** — Both regime 1 (sweep migration) and regime 2 (past-cutoff held-back) are tested. Second migration call correctly asserts `None`.

### ⚠️ Protocol-critical — requires human sign-off

This PR modifies VTXO batch input selection, forfeit path exclusion, boarding output filtering, and balance classification. These are all protocol-critical paths where bugs lose real money. The code is well-structured and I found no fund-loss vectors, but per project policy this requires explicit human review before merge.

**Review by ghost — APPROVED** [review]
## 🔒 Arkana Follow-up Review — New Commits `c435842`, `7b7a6cf`

Both MEDIUM issues from my previous review are resolved:

### ✅ Time source unification (`c435842`)
- `jiff::Timestamp::now()` replaced with `super::unix_now()` in `batch.rs`
- `now_secs` captured once at top of `fetch_commitment_transaction_inputs`, reused for both boarding outputs and VTXOs — no more dual-capture window
- Duration conversion done inline with `std::time::Duration::from_secs(now_secs as u64)` — correct

### ✅ TOCTOU fix (`7b7a6cf`)
- New `settle_at(now, rng)` takes timestamp parameter; `settle()` is a thin wrapper calling `settle_at(unix_now(), rng)`
- `migrate_deprecated_signer_vtxos()` now passes its captured `now` through to `settle_at` → `fetch_commitment_transaction_inputs(now)` — single time snapshot across the entire migration
- All 7 other call sites of `fetch_commitment_transaction_inputs` updated to pass `super::unix_now()` at their own call site — no missed callers
- `fee_estimation.rs` callers also updated (2 sites) — complete

### Remaining LOW items from previous review (non-blocking)
- `cutoff_date == 0` sentinel docs — still nice-to-have but not blocking
- Manual VTXO selection paths — still don't filter past-cutoff, but server rejects gracefully

**Approving from my side. Protocol-critical flag still applies — this still needs human sign-off before merge (VTXO batch input selection, forfeit path exclusion, balance classification).**

**Review by ghost — APPROVED** [review]
## 🔍 Arkana Follow-up Review — New Commit `4f1c7aae`

**Change**: `migrate_deprecated_signer_vtxos` now filters inputs to only include deprecated-signer VTXOs/boarding outputs, instead of calling `settle_at` (which would consolidate *all* VTXOs including current-signer ones).

### Assessment: ✅ Correct fix, well-tested

**Logic review** — the new flow is:
1. Collect `deprecated_vtxo_outpoints` from `spendable_offchain()` (pre-cutoff deprecated signers only)
2. Check `migratable_boarding` from boarding outputs
3. Call `fetch_commitment_transaction_inputs(now)` with same timestamp snapshot (TOCTOU fix preserved ✓)
4. **Post-filter**: keep only boarding inputs under deprecated signers + VTXO inputs whose outpoints are in the pre-computed set
5. Recompute `total_amount` from filtered inputs
6. Guard against empty inputs after filtering (handles edge case where `fetch_commitment_transaction_inputs` already excluded some)
7. Call `join_next_batch` directly instead of `settle_at`

This is sound. The outpoint-based filtering is the right approach — it's a stable identifier that can't drift between the pre-check and the input fetch.

**One nit** — the doc comment (lines ~1111-1117) is now stale. It still says:
> "Internally this calls `Self::settle`, which settles **all** unspent VTXOs and boarding outputs in a single batch — not just the deprecated-signer ones."

That's exactly what this commit fixed. The comment should be updated to reflect the new selective behavior.

**E2E test** (`e2e_signer_rotation_only_deprecated_vtxos_migrated`): Good coverage. Creates VTXOs under two signer generations, migrates, and asserts current-signer outpoints survive unchanged. This directly validates the fix.

**Protocol safety**: No concerns. This change is strictly *more* conservative — it moves fewer VTXOs, reducing unnecessary churn. The cutoff logic and TOCTOU protections from prior commits are preserved.

Approving. The stale docstring is low-priority — can be fixed in a follow-up.

🔒 *This PR modifies VTXO handling and round participation logic. Human approval is still required before merge.*

File: ark-client/src/lib.rs [hunk 1]
```diff
@@ -392,6 +402,22 @@ pub trait Blockchain {
...
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("valid duration")
+            .as_secs() as i64
+    }
+
+    #[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+    {
+        (js_sys::Date::now() / 1000.0) as i64
+    }
+}
```
Comment (aruokhai) [inline ark-client/src/lib.rs:419]: I do not think wasm is currently being supported, feel free to drop , @bonomat  am I wrong ?

File: ark-core/src/server.rs [hunk 2]
```diff
@@ -776,3 +799,137 @@ pub fn parse_fee_amount(amount_str: Option<String>) -> Amount {
...
+        let keys: Vec<_> = info.all_server_keys().collect();
+        assert_eq!(keys, vec![xonly(PK_A), xonly(PK_B), xonly(PK_C)]);
+    }
+
+    #[test]
+    fn all_server_keys_current_is_always_first() {
+        let info = make_info(PK_C, vec![(PK_A, 500), (PK_B, 600)]);
+        let keys: Vec<_> = info.all_server_keys().collect();
+        assert_eq!(keys[0], xonly(PK_C));
+    }
```
Comment (aruokhai) [inline ark-core/src/server.rs:882]: why is this necessary ?

File: ark-rest/src/conversions.rs [hunk 3]
```diff
@@ -117,11 +117,12 @@ impl TryFrom<crate::models::DeprecatedSigner> for ark_core::server::DeprecatedSi
...
-            ConversionError("Missing cutoff_date in deprecated signer".to_string())
-        })?;
-        let cutoff_date = i64::from_str(&cutoff_date_str)
-            .map_err(|e| ConversionError(format!("Could not parse cutoff_date: {e:#}")))?;
+        // A missing cutoffDate means "rotate immediately" (same as 0).
+        let cutoff_date = match value.cutoff_date {
+            Some(s) => i64::from_str(&s)
+                .map_err(|e| ConversionError(format!("Could not parse cutoff_date: {e:#}")))?,
+            None => 0,
+        };
```
Comment (aruokhai) [inline ark-rest/src/conversions.rs:125]: Can we Keep this as Option<i64>, collapsing None to 0 erases the absent-vs-epoch distinction.

File: ark-client/src/lib.rs [hunk 4]
```diff
@@ -1034,10 +1106,110 @@ where
...
+
+        let now = unix_now();
+        let (vtxo_list, script_map) = self.list_vtxos().await?;
+
+        let is_pre_cutoff_deprecated = |server_pk: XOnlyPublicKey| {
+            self.server_info.deprecated_signers.iter().any(|ds| {
+                ds.pk.x_only_public_key().0 == server_pk
+                    && (ds.cutoff_date == 0 || ds.cutoff_date > now)
+            })
+        };
```
Comment (aruokhai) [inline ark-client/src/lib.rs:1143]: If we retain cutoff as an option, we can do this

```rust
let is_pre_cutoff_deprecated = |server_pk: XOnlyPublicKey| {
    self.server_info.deprecated_signers.iter().any(|ds| {
        ds.pk.x_only_public_key().0 == server_pk
            && ds.cutoff_date.is_none_or(|cutoff| cutoff > now)
    })
};
```

**Comment by d4rp4t** [comment]
replaced by #244
